src/script.c

```diff
@@ -141,15 +145,234 @@ void environment_exit(environment_t *env) {
 	free(env->entries);
 }
 
+static int run_command(const char *scriptname) {
+	char buf[8192];
+
+	if(scriptinterpreter) {
+		snprintf(buf, sizeof(buf), "%s \"%s\"", scriptinterpreter, scriptname);
+	} else {
+		snprintf(buf, sizeof(buf), "\"%s\"", scriptname);
+	}
+
+	return system(buf);
+}
+
+static bool build_script_name(char *buf, size_t len, const char *name) {
+	if(!name) {
+		return false;
+	}
+
+	// If name contains a forward slash, make sure it is the only one, and
+	// it forms a separator between the 'hosts' subdirectory and a file inside it.
+	// This should prevent attackers from using names like `/ysr/bin/python` or `../../python3`.
+	const char *slash = strchr(name, '/');
+
+	if(slash) {
+		if(strncmp("hosts/", name, sizeof("hosts/") - 1) || strchr(slash + 1, '/')) {
+			return false;
+		}
+	}
+
+	int wrote = snprintf(buf, len, "%s" SLASH "%s%s", confbase, name, scriptextension);
+	return wrote > 0 && (size_t)wrote <= len;
+}
+
+#ifdef HAVE_SANDBOX
+struct {
+	pid_t pid;
+	int sock;
+	bool stopped;
+} worker;
+
+static void cleanup_script_worker(void) {
+	if(worker.sock) {
+		logger(DEBUG_ALWAYS, LOG_INFO, "Waiting for script runner to exit");
+		close(worker.sock);
+		waitpid(worker.pid, NULL, 0);
+		worker.stopped = true;
+		worker.sock = 0;
+	}
+}
+
+void set_script_worker(pid_t pid, int sock) {
+	assert(!worker.pid && !worker.sock && !worker.stopped);
+	worker.pid = pid;
+	worker.sock = sock;
+	atexit(cleanup_script_worker);
+}
+
+static const char *known_vars[] = {
+	"DEBUG",
+	"DEVICE",
+	"INTERFACE",
+	"INVITATION_FILE",
+	"INVITATION_URL",
+	"NAME",
+	"NETNAME",
+	"NODE",
+	"REMOTEADDRESS",
+	"REMOTEPORT",
+	"SUBNET",
+	"WEIGHT",
+	NULL,
+};
+
+static bool is_known_var(const char *var) {
+	for(const char **known = known_vars; *known; ++known) {
+		if(!strcmp(*known, var)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static bool read_unix(int fd, void *data, size_t len) {
+	ssize_t res = read(fd, data, len);
+	return res > 0 && (size_t)res == len;
+}
+
+static bool write_unix(int fd, const void *data, size_t len) {
+	ssize_t res = write(fd, data, len);
+	return res > 0 && (size_t)res == len;
+}
```

Comment:
Most UNIX socket implementations are always reliable and in-order, whether it is stream or datagram oriented, but what is not guaranteed on a `SOCK_STREAM` socket is that a `write()` will complete in its entirety, nor that every `read()` will return exactly what a single `write()` sent. Either use `SOCK_DGRAM`, or even better, `SOCK_SEQPACKET`. You then also don't have to encode the length in a header.

Reply:
FWIW, `SEQPACKET` is not available on NetBSD and macOS. NetBSD doesn't seem to have any sandboxing capabilities, but macOS does. I'm not planning on implementing it since I don't have access to macOS and don't care for it much, but if someone does, that's something to think about.

It's a lot nicer than sending lengths manually, so I left `SEQPACKET` for now.